pkg/metrics/metrics.go

```diff
@@ -0,0 +1,275 @@
+	incWebhooksWithLabelValues(*webhookRequestPassCount, isValidating, obj)
+}
+
+func IncWebhookRequestRejectCount(isValidating bool, obj runtime.Object) {
+	incWebhooksWithLabelValues(*webhookRequestRejectCount, isValidating, obj)
+}
+
+func IncWebhookRequestPanicCount(isValidating bool, obj runtime.Object) {
+	incWebhooksWithLabelValues(*webhookRequestPanicCount, isValidating, obj)
+}
+
+// Get webhook metric count
+
+func GetWebhookRequestPanicCountWith(isValidating bool, obj runtime.Object) int {
+	return getWebhooksWithLabelValues(*webhookRequestPanicCount, isValidating, obj)
+}
+
+func GetWebhookRequestRejectCount(isValidating bool, obj runtime.Object) int {
+	return getWebhooksWithLabelValues(*webhookRequestRejectCount, isValidating, obj)
+}
+
+func GetWebhookRequestCount(isValidating bool, obj runtime.Object) int {
+	return getWebhooksWithLabelValues(*webhookRequestCount, isValidating, obj)
+}
+
+func GetWebhookRequestPassCount(isValidating bool, obj runtime.Object) int {
+	return getWebhooksWithLabelValues(*webhookRequestPassCount, isValidating, obj)
+}
+
+// Increment/update controller metric
+
+// incControllersWithLabelValues increments the controller metric counter 'metric' by one.
+func incControllersWithLabelValues(metric prometheus.CounterVec, req reconcile.Request, rw risingwavev1alpha1.RisingWave) {
+	gvk := rw.GroupVersionKind()
+	metric.WithLabelValues(gvk.Group, gvk.Version, gvk.Kind, req.Namespace, req.Name).Inc()
+}
+
+func IncControllerReconcileCount(req reconcile.Request, rw risingwavev1alpha1.RisingWave) {
+	incControllersWithLabelValues(*controllerReconcileCount, req, rw)
+}
+
+func IncControllerReconcilePanicCount(req reconcile.Request, rw risingwavev1alpha1.RisingWave) {
+	incControllersWithLabelValues(*controllerReconcilePanicCount, req, rw)
+}
+
+func IncControllerReconcileRequeueCount(req reconcile.Request, rw risingwavev1alpha1.RisingWave) {
+	incControllersWithLabelValues(*controllerReconcileRequeueCount, req, rw)
+}
+
+func UpdateControllerReconcileRequeueAfter(time_ms int64, req reconcile.Request, rw risingwavev1alpha1.RisingWave) {
+	gvk := rw.GroupVersionKind()
+	controllerReconcileRequeueAfter.WithLabelValues(gvk.Group, gvk.Version,
+		gvk.Kind, req.Namespace, req.Name).Observe(float64(time_ms))
+}
+
+func IncControllerReconcileRequeueErrorCount(req reconcile.Request, rw risingwavev1alpha1.RisingWave) {
+	incControllersWithLabelValues(*controllerReconcileRequeueErrorCount, req, rw)
+}
+
+func UpdateControllerReconcileDuration(time_ms int64, risingwave risingwavev1alpha1.RisingWave, webhookName string) {
+	gvk := risingwave.GroupVersionKind()
+	controllerReconcileDuration.WithLabelValues(webhookName, gvk.Group, gvk.Version,
+		gvk.Kind, risingwave.Namespace, risingwave.Name).Observe(float64(time_ms))
+}
+
+// ResetMetrics resets all metrics. Use for testing only.
+func ResetMetrics() {
+	var m io_prometheus_client.Metric
+	m.Reset()
+	ReceivingMetricsFromOperator.Write(&m)
```

Comment:
In this function I want to reset all the metrics. Otherwise the tests will not work. In the tests we are asserting that a metric has some value, but if a different test was called before the value of the metric was changed. 

You can reset `prometheus.CounterVec` via `Reset()`, but for `prometheus.Counter` you need to write a value to the counter. I am writing an empty metric to `ReceivingMetricsFromOperator.Write` here. 

I will abbreviate the snippet to `ReceivingMetricsFromOperator.Write(&io_prometheus_client.Metric{})`. Maybe that is a bit clearer